### PR TITLE
4117: Added Md5String() method and updated tests.

### LINKED ISSUE #4117: blob.getMd5() doesn't work as expected
When I get any blob MD5 checksum using java API, the result is a Base64 string, that, when decoded, is not a MD5 string.

Expected result: the file MD5 checksum string.

Can anyone explain me why getMd5() returns a base64 of a MD5 instead of the MD5 itself?

#### Environment details

- OS: Linux
- Java version: 1.8.0_172
- google-cloud-java version(s): google-cloud-storage:1.53.0

#### Steps to reproduce

1. Save a blob using bucket.create(completeFilePath, fileByteArray)
2. Get the file metadata using storage.get(blobId)
3. Get the file MD5 using blob.getMd5()

### PR DESCRIPTION
**Pull request by andrey-qlogic**
Fixes #4117

### CODE REVIEW

**Comment by andrey-qlogic** [comment]
@frankyn, would you please take a look at this PR.

**Comment by andrey-qlogic** [comment]
Added Crc32cString() method and updated tests. Added to snippets.

**Comment by frankyn** [comment]
Hi @andrey-qlogic, Thanks for the update. I chatted with the GCS team about this design. It sounds good to them but suggested adding a Hex Hash String to GCS base64 encoded string for setCrc32c() method to the `Blob class`.

The naming convention would help: @googleapis/yoshi-java do you have input on Java community method names? 

WDYT about the following names?
```java
getMd5ToHexString()
setMd5FromHexString()

getCrc32cToHexString()
setCrc32cFromHexString()
```

**Comment by andrey-qlogic** [comment]
Renamed new methods using naming convention getMd5ToHexString(), getCrc32cToHexString().
I believe that setter methods are not required as those new methods return a string representation md5 and cr32 that set in another method.

**Comment by frankyn** [comment]
The other method [setMd5](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java#L422) for example expects the base64 encoded string of an Md5 hash. It would be nice to have a similar helper method to getting the md5 hash as a base64 decoded md5 hex string.

**Comment by andrey-qlogic** [comment]
@frankyn, that is unclear, are you suggesting to add a new setter method to class BuilderImpl 
 like 
`public Builder setMd5FromHexString(String md5HexString)`
that will do something like 
`byte[] decoded = BaseEncoding.base64().decode(md5HexString);`

and assign it to md5 ?

Otherwise the method getMd5ToHexString is implepented for BlobInfo class in https://github.com/googleapis/google-cloud-java/pull/4397/commits/85588a3c990f2a080ed38ee4555711ca33f1522d

I would appreciate for your input

**Comment by andrey-qlogic** [comment]
Added methods to set Md2 and Crc32s from hex strings.

**Comment by andrey-qlogic** [comment]
there are code format issues that will be fixed with #4437

File: google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java [hunk 1]
```diff
@@ -423,12 +441,37 @@ public Builder setMd5(String md5) {
...
 
+    public Builder setMd5FromHexString(String md5HexString) {
+      if (md5HexString == null) {
+        return this;
+      }
+      if (md5HexString.startsWith("0x")) {
+        md5HexString = md5HexString.replaceFirst("0x", "");
+      }
+      byte[] bytes = new BigInteger(md5HexString, 16).toByteArray();
+      this.md5 = "0x" + BaseEncoding.base64().omitPadding().encode(bytes);
```
Comment (frankyn) [inline google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java]: Why does this prepend `0x`? 

Could you add an integration that exercises the `setMd5/Crc32cFromHexString` methods?

**Comment by andrey-qlogic** [comment]
@frankyn, I realized that Md5 sometimes used with 0x prefix that can be easily removed at BlobInfo class.

**Comment by frankyn** [comment]
For this PR, @andrey-qlogic could you skip supporting the removal of "0x" for now? 
I'm leaning towards supporting that incrementally instead of in the first iteration.

Thank you, apologies for constant delays on these reviews.

**Comment by andrey-qlogic** [comment]
@frankyn, do you want me to drop the whole last commit [https://github.com/googleapis/google-cloud-java/pull/4397/commits/7f19e33c3516c119f1a2775d0eb0fd07533a06c7](https://github.com/googleapis/google-cloud-java/pull/4397/commits/7f19e33c3516c119f1a2775d0eb0fd07533a06c7) ?

**Comment by frankyn** [comment]
Yes except the integration test.

**Comment by andrey-qlogic** [comment]
Reverted back the commit [https://github.com/googleapis/google-cloud-java/commit/7f19e33c3516c119f1a2775d0eb0fd07533a06c7](https://github.com/googleapis/google-cloud-java/commit/7f19e33c3516c119f1a2775d0eb0fd07533a06c7) except integration tests.

File: google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java (hunk 1 above)
Comment (frankyn) [inline google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java]: Please remove prepending "0x".

File: google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java [hunk 2]
```diff
@@ -423,12 +441,37 @@ public Builder setMd5(String md5) {
       return this;
     }
 
+    public Builder setMd5FromHexString(String md5HexString) {
+      if (md5HexString == null) {
+        return this;
+      }
+      if (md5HexString.startsWith("0x")) {
+        md5HexString = md5HexString.replaceFirst("0x", "");
+      }
```
Comment (frankyn) [inline google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java]: Remove code that removes "0x" from the hex string.

File: google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java (hunk 1 above)
Comment (andrey-qlogic) [reply google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java]: @frankyn, do you mean the change for both Md5 and Crc32c, correct 
What about unit tests values? https://github.com/googleapis/google-cloud-java/blob/7de46b4ce194f4c7d7094dd1d7dec4f4a85ad363/google-cloud-clients/google-cloud-storage/src/test/java/com/google/cloud/storage/BlobInfoTest.java#L55

We allow user to enter the value with 0x and without.

**Comment by andrey-qlogic** [comment]
@frankyn, do you mean the change for both Md5 and Crc32c, correct
What about unit tests values?

https://github.com/googleapis/google-cloud-java/blob/7de46b4ce194f4c7d7094dd1d7dec4f4a85ad363/google-cloud-clients/google-cloud-storage/src/test/java/com/google/cloud/storage/BlobInfoTest.java#L55

File: google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java (hunk 2 above)
Comment (andrey-qlogic) [reply google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobInfo.java]: Removed code that removes 0x from hex string and code that adds prepending 0x. OmitPadding removed

**Comment by andrey-qlogic** [comment]
Removed code that removes 0x from hex string and code that adds prepending 0x. OmitPadding removed

File: google-cloud-clients/google-cloud-storage/src/test/java/com/google/cloud/storage/BlobTest.java [hunk 3]
```diff
@@ -71,11 +71,13 @@
   private static final String CONTENT_ENCODING = "UTF-8";
   private static final String CONTENT_LANGUAGE = "En";
   private static final String CRC32 = "0xFF00";
```
Comment (frankyn) [inline google-cloud-clients/google-cloud-storage/src/test/java/com/google/cloud/storage/BlobTest.java]: I understand the confusion now. Sorry I didn't see this before, thanks for linking to it.

The hex values used for `CRC32` and `MD5` are not valid and only used for unit tests. I recommend replacing these values out with the base64 encoded crc32c and md5 values.

**Comment by andrey-qlogic** [comment]
@frankyn , I realized that new BigInteger(md5HexString, 16).toByteArray() is not safe with leading zeros. I changed to DatatypeConverter.parseHexBinary(md5HexString). It works fine. If needed we can make an utility method instead.

**Comment by andrey-qlogic** [comment]
thanks for the approval

**Pull request merged**